### PR TITLE
fix to input to aia_color_table 

### LINKED ISSUE #3480: Input to aia_color_table should be unitful
### Description

`aia_color_table`, the function for selecting an AIA color map, takes in a selection for the AIA wavelength. However,  the accepted input types are either a string or integer. It should be an `astropy.units.Quantity` as the channel is a wavelength and a wavelength is a physical quantity

### PR DESCRIPTION
**Pull request by mridullpandey**
### Description

Fixes #3480

### CODE REVIEW

File: sunpy/visualization/colormaps/color_tables.py [hunk 1]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
+b0 = np.array(padfr(_mkx(3, range(51, 256, 51), 4), 256))*u.angstrom
 
-c0 = np.arange(256, dtype='f')
-c1 = (np.sqrt(c0) * np.sqrt(255.0)).astype('f')
-c2 = (np.arange(256)**2 / 255.0).astype('f')
-c3 = ((c1 + c2 / 2.0) * 255.0 / (c1.max() + c2.max() / 2.0)).astype('f')
+c0 = np.arange(256, dtype='f')*u.angstrom
+c1 = (np.sqrt(c0) * np.sqrt(255.0)).astype('f')*u.angstrom
+c2 = (np.arange(256)**2 / 255.0).astype('f')*u.angstrom
+c3 = ((c1 + c2 / 2.0) * 255.0 / (c1.max() + c2.max() / 2.0)).astype('f')*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]: You don't want to change the table data to be units. These are used by matplotlib to color in images.

File: sunpy/visualization/colormaps/color_tables.py [hunk 2]
```diff
@@ -81,7 +82,7 @@ def paden(lst, len_, pad=0):
 }
 
 
-def aia_color_table(wavelength):
+def aia_color_table(wavelength: u.angstrom):
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
@u.quantity_input
def aia_color_table(wavelength: u.angstrom):
```

As we require from: https://docs.sunpy.org/en/latest/dev_guide/units_quantities.html

File: sunpy/visualization/colormaps/color_tables.py (hunk 1 above)
Comment (mridullpandey) [reply sunpy/visualization/colormaps/color_tables.py]: yes..
I'm sorry.. i failed ,any hints please!

Comment (nabobalis) [reply sunpy/visualization/colormaps/color_tables.py]: That's fine! This is all new to you, we can take this slowly.

**Comment by nabobalis** [comment]
We also need to change: https://github.com/sunpy/sunpy/pull/3499/files#diff-d4ec3850567004e6e4a003c7ea5001ecR94 to not say `int` anymore but `~astropy.units.quantity.Quantity`.

The question is now, there are probably tests and other lines of code we need to change, but we can tackle it in order.

File: sunpy/visualization/colormaps/color_tables.py (hunk 1 above)
Comment (mridullpandey) [reply sunpy/visualization/colormaps/color_tables.py]: okay thank you so much...

File: sunpy/visualization/colormaps/color_tables.py [hunk 3]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
         205, 207, 208, 210, 211, 212, 214, 215, 217, 218, 220, 221, 223, 224,
         226, 227, 228, 230, 231, 233, 234, 236, 237, 239, 240, 241, 243, 244,
         246, 247, 249, 250, 252, 253
-    ], 256, 255))
+    ], 256, 255))*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
    ], 256, 255))
```

File: sunpy/visualization/colormaps/color_tables.py [hunk 4]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
 
-g0 = np.array(padfr(_mkx(1, range(17, 256, 17), 2), 256))
-b0 = np.array(padfr(_mkx(3, range(51, 256, 51), 4), 256))
+g0 = np.array(padfr(_mkx(1, range(17, 256, 17), 2), 256))*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
g0 = np.array(padfr(_mkx(1, range(17, 256, 17), 2), 256))
```

File: sunpy/visualization/colormaps/color_tables.py [hunk 5]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
+b0 = np.array(padfr(_mkx(3, range(51, 256, 51), 4), 256))*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
b0 = np.array(padfr(_mkx(3, range(51, 256, 51), 4), 256))
```

File: sunpy/visualization/colormaps/color_tables.py [hunk 6]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
 
-c0 = np.arange(256, dtype='f')
-c1 = (np.sqrt(c0) * np.sqrt(255.0)).astype('f')
-c2 = (np.arange(256)**2 / 255.0).astype('f')
-c3 = ((c1 + c2 / 2.0) * 255.0 / (c1.max() + c2.max() / 2.0)).astype('f')
+c0 = np.arange(256, dtype='f')*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
c0 = np.arange(256, dtype='f')
```

File: sunpy/visualization/colormaps/color_tables.py [hunk 7]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
+c1 = (np.sqrt(c0) * np.sqrt(255.0)).astype('f')*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
c1 = (np.sqrt(c0) * np.sqrt(255.0)).astype('f')
```

File: sunpy/visualization/colormaps/color_tables.py [hunk 8]
```diff
@@ -57,15 +58,15 @@ def paden(lst, len_, pad=0):
...
+c2 = (np.arange(256)**2 / 255.0).astype('f')*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
c2 = (np.arange(256)**2 / 255.0).astype('f')
```

File: sunpy/visualization/colormaps/color_tables.py (hunk 1 above)
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
c3 = ((c1 + c2 / 2.0) * 255.0 / (c1.max() + c2.max() / 2.0)).astype('f')
```

**Comment by Cadair** [comment]
Thanks for the PR @mridullpandey :smile:

File: sunpy/visualization/colormaps/color_tables.py (hunk 2 above)
Comment (mridullpandey) [reply sunpy/visualization/colormaps/color_tables.py]: okay i got it i guess..thanks:-)

**Comment by nabobalis** [comment]
Since this is an API change, what's the policy here @Cadair ?

**Comment by nabobalis** [comment]
@mridullpandey as I make suggestions, you can use the GitHub UI to accept them as a collection and commit those changes without having to do them manually.

File: sunpy/visualization/colormaps/color_tables.py (hunk 2 above)
Comment (mridullpandey) [reply sunpy/visualization/colormaps/color_tables.py]: as i don't import astropy.units as u first it errors as  'u' is not recognised..

Comment (nabobalis) [reply sunpy/visualization/colormaps/color_tables.py]: Well the import ideally should be at the top of the file. So it would work.

**Comment by nabobalis** [comment]
Now the next step is to rewrite how we index the dictionary since I don't think astropy.units are allowed to be keys (maybe they can be?).

**Comment by mridullpandey** [comment]
okay @nabobalis

**Comment by nabobalis** [comment]
You'll also need to find any places were that function is called and replace the wavelength argument   with an astropy.unit input otherwise it will just error (which is one reason the tests are all failing at the bottom).

**Comment by mridullpandey** [comment]
aia_wave_dict = {
    1600: (c3, c3, c2)*u.angstrom,
    1700: (c1, c0, c0)*u.angstrom,
    4500: (c0, c0, b0 / 2.0)*u.angstrom,
    94: (c2, c3, c0)*u.angstrom,
    131: (g0, r0, r0)*u.angstrom,
    171: (r0, c0, b0)*u.angstrom,
    193: (c1, c0, c2)*u.angstrom,
    211: (c1, c0, c3)*u.angstrom,
    304: (r0, g0, b0)*u.angstrom,
    335: (c2, c0, c1)*u.angstrom
}
how would be appying the above code  to wave_dict ..
yet it runs fine in console..

**Comment by mridullpandey** [comment]
will do!! 
@nabobalis

**Comment by nabobalis** [comment]
> aia_wave_dict = {
> 1600: (c3, c3, c2)*u.angstrom,
> 1700: (c1, c0, c0)*u.angstrom,
> 4500: (c0, c0, b0 / 2.0)*u.angstrom,
> 94: (c2, c3, c0)*u.angstrom,
> 131: (g0, r0, r0)*u.angstrom,
> 171: (r0, c0, b0)*u.angstrom,
> 193: (c1, c0, c2)*u.angstrom,
> 211: (c1, c0, c3)*u.angstrom,
> 304: (r0, g0, b0)*u.angstrom,
> 335: (c2, c0, c1)*u.angstrom
> }
> how would be appying the above code to wave_dict ..
> yet it runs fine in console..

If that works, we can do that.

**Comment by mridullpandey** [comment]
> If that works, we can do that.

okay

File: sunpy/visualization/colormaps/color_tables.py [hunk 9]
```diff
@@ -69,16 +69,16 @@ def paden(lst, len_, pad=0):
...
+    1600: (c3, c3, c2)*u.angstrom,
+    1700: (c1, c0, c0)*u.angstrom,
+    4500: (c0, c0, b0 / 2.0)*u.angstrom,
+    94: (c2, c3, c0)*u.angstrom,
+    131: (g0, r0, r0)*u.angstrom,
+    171: (r0, c0, b0)*u.angstrom,
+    193: (c1, c0, c2)*u.angstrom,
+    211: (c1, c0, c3)*u.angstrom,
+    304: (r0, g0, b0)*u.angstrom,
+    335: (c2, c0, c1)*u.angstrom
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]: Ah sorry I should have read the example clearer. You want these on the keys and not the values of the dictionary.

File: sunpy/visualization/colormaps/color_tables.py [hunk 10]
```diff
@@ -68,20 +69,22 @@ def paden(lst, len_, pad=0):
...
-    1700: (c1, c0, c0),
-    4500: (c0, c0, b0 / 2.0),
-    94: (c2, c3, c0),
-    131: (g0, r0, r0),
-    171: (r0, c0, b0),
-    193: (c1, c0, c2),
-    211: (c1, c0, c3),
-    304: (r0, g0, b0),
-    335: (c2, c0, c1)
+    1600: (c3, c3, c2)*u.angstrom,
```
Comment (nabobalis) [inline sunpy/visualization/colormaps/color_tables.py]:
```suggestion
    1600*u.angstrom: (c3, c3, c2),
```
Like that and then in theory we can index the dict with astropy units. Will need to test this to make sure.

**Comment by Cadair** [comment]
> Since this is an API change, what's the policy here @Cadair ?

I think this is going to be pretty difficult to deprecate, so we should just do it as a breaking change in a major release.

File: sunpy/visualization/colormaps/color_tables.py (hunk 10 above)
Comment (wtbarnes) [reply sunpy/visualization/colormaps/color_tables.py]: You can in fact use quantities as dictionary keys 🚀

**Comment by mridullpandey** [comment]
hi @nabobalis ,
Due to semester exams i would not be able to active on this,as soon as i get free i'll be right back.
by the way it was lovely exeperience working with you. Loved opensoarce a lot .
thanks :-)

**Comment by nabobalis** [comment]
> hi @nabobalis ,
> Due to semester exams i would not be able to active on this,as soon as i get free i'll be right back.
> by the way it was lovely exeperience working with you. Loved opensoarce a lot .
> thanks :-)

That is perfectly fine! Good luck with your exams!

**Comment by mridullpandey** [comment]
hi greetings!!
sorry for the break,i'm back can i proceed further into it..

**Comment by nabobalis** [comment]
Welcome back. Take as long as you need to get back up to speed.

**Comment by mridullpandey** [comment]
hi can i make a new pr for the same...

**Comment by nabobalis** [comment]
Up to you.

**Pull request closed without merging**